### ISSUE TITLE
Address Sanitizer, Buffer Overflow?

### ISSUE THREAD

**Issue by Bluezen** [lucasmedeirosleite/EasyMapping#155]
Hello,
I stumbled upon a possible buffer overflow in `propertyNameIsScalar:fromObject:` of `EKPropertyHelper` at the following line [L46](https://github.com/lucasmedeirosleite/EasyMapping/blob/master/EasyMapping/EKPropertyHelper.m#L46):
```objc
return (type.length == 1) && (NSNotFound != [@(scalarTypes) rangeOfString:type].location);
```

The **Address Sanitizer** tool seems to complain about the boxed C string. If I got it right from the [clang doc](https://clang.llvm.org/docs/ObjectiveCLiterals.html#boxed-c-strings) it's waiting for a \0 terminated string...

Without touching `scalarTypes` we could search for `type` using C memchr:
```objc
return (type.length == 1) && memchr(scalarTypes, type.UTF8String[0], sizeof(scalarTypes));
```
This way the Address Sanitizer stopped complaining. I have the modification in a [commit](https://github.com/Bluezen/EasyMapping/commit/030a184feaac585a4f7bea12c7b6b4978068a9e7#diff-bd41a82ec8ad18eb0f62ff9448ce0418L46)  if you want me to make a pull request.

The stacktrace:
```
=================================================================
==18506==ERROR: AddressSanitizer: global-buffer-overflow on address 0x000102afe62e at pc 0x00010189dfd9 bp 0x7fff5eb9cf60 sp 0x7fff5eb9c708
READ of size 15 at 0x000102afe62e thread T0
    #0 0x10189dfd8 in wrap_strlen (libclang_rt.asan_iossim_dynamic.dylib:x86_64+0x10fd8)
    #1 0x1050a0245 in +[NSString stringWithUTF8String:] (Foundation:x86_64+0x6245)
    #2 0x102ae6b20 in +[EKPropertyHelper propertyNameIsScalar:fromObject:] EKPropertyHelper.m:47
    #3 0x102ae85de in +[EKPropertyHelper setProperty:onObject:fromRepresentation:respectPropertyType:ignoreMissingFields:] EKPropertyHelper.m:104
    #4 0x102ad286b in __62+[EKMapper fillObject:fromExternalRepresentation:withMapping:]_block_invoke EKMapper.m:50
    #5 0x105af65fe in __65-[__NSDictionaryM enumerateKeysAndObjectsWithOptions:usingBlock:]_block_invoke (CoreFoundation:x86_64+0x7a5fe)
    #6 0x105af64f9 in -[__NSDictionaryM enumerateKeysAndObjectsWithOptions:usingBlock:] (CoreFoundation:x86_64+0x7a4f9)
    #7 0x102ad03a9 in +[EKMapper fillObject:fromExternalRepresentation:withMapping:] EKMapper.m:49
    #8 0x102acf963 in +[EKMapper objectFromExternalRepresentation:withMapping:] EKMapper.m:42
    #9 0x102ad3f28 in +[EKMapper arrayOfObjectsFromExternalRepresentation:withMapping:] EKMapper.m:144
    #10 0x1012e6d85 in __71+ *redacted*
    #11 0x1011a1cdc in __84-[AFHTTPSessionManager dataTaskWithHTTPMethod:URLString:parameters:success:failure:]_block_invoke.93 AFHTTPSessionManager.m:250
    #12 0x10122aa48 in __72-[AFURLSessionManagerTaskDelegate URLSession:task:didCompleteWithError:]_block_invoke_2.74 AFURLSessionManager.m:194
    #13 0x1018dd42e in __wrap_dispatch_group_async_block_invoke (libclang_rt.asan_iossim_dynamic.dylib:x86_64+0x5042e)
    #14 0x1077554a5 in _dispatch_call_block_and_release (libdispatch.dylib:x86_64+0x24a5)
    #15 0x10777e05b in _dispatch_client_callout (libdispatch.dylib:x86_64+0x2b05b)
    #16 0x10775f582 in _dispatch_main_queue_callback_4CF (libdispatch.dylib:x86_64+0xc582)
    #17 0x105b32908 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ (CoreFoundation:x86_64+0xb6908)
    #18 0x105af8ae3 in __CFRunLoopRun (CoreFoundation:x86_64+0x7cae3)
    #19 0x105af8015 in CFRunLoopRunSpecific (CoreFoundation:x86_64+0x7c015)
    #20 0x108ca3a23 in GSEventRunModal (GraphicsServices:x86_64+0xca23)
    #21 0x1033570d3 in UIApplicationMain (UIKit:x86_64+0x280d3)
    #22 0x10142135e in main main.m:15
    #23 0x1077ca65c in start (libdyld.dylib:x86_64+0x465c)`

0x000102afe62e is located 0 bytes to the right of global variable 'scalarTypes' defined in '/$$myapp$$/Pods/EasyMapping/EasyMapping/EKPropertyHelper.m:28:19' (0x102afe620) of size 14
SUMMARY: AddressSanitizer: global-buffer-overflow (libclang_rt.asan_iossim_dynamic.dylib:x86_64+0x10fd8) in wrap_strlen
Shadow bytes around the buggy address:
  0x10002055fc70: f9 f9 f9 f9 00 00 00 00 00 01 f9 f9 f9 f9 f9 f9
  0x10002055fc80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fc90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fca0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fcb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10002055fcc0: 00 00 00 00 00[06]f9 f9 f9 f9 f9 f9 00 00 00 00
  0x10002055fcd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fce0: 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9 f9 f9 f9
  0x10002055fcf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fd00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002055fd10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==18506==ABORTING
```

**Comment by DenHeadless** [comment]
Hey!

Sure, that would be great. Even if address sanitizer is catching a false positive, we should consider this a warning and write this part in different way.

**Comment by DenHeadless** [comment]
I've cherry-picked this fix into different branch, had a lot of moving parts there)

I released this fix as a part of [0.21.0 release](https://github.com/lucasmedeirosleite/EasyMapping/releases/tag/0.21.0), thanks for contributing!

**Comment by Bluezen** [comment]
You are welcome! Thanks for your time on this great project!
